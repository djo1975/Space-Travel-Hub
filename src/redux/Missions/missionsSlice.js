import { createSlice, createAsyncThunk, createAction } from '@reduxjs/toolkit';

export const fetchMissions = createAsyncThunk('missions/fetchMissions', async () => {
  const response = await fetch('https://api.spacexdata.com/v3/missions');
  const data = await response.json();
  return data;
});
export const joinMission = createAction('missions/joinMission');

const missionsSlice = createSlice({
  name: 'missions',
  initialState: [],
  reducers: {
    joinMission: (state, action) => {
      const id = action.payload;
      return state.map((mission) => {
        if (mission.mission_id !== id) {
          return mission;
        }
        const newMission = { ...mission, joined: true };
        return newMission;
      });
    },
    leavingMission: (state, action) => {
      const id = action.payload;
      return state.map((mission) => {
        if (mission.mission_id !== id) {
          return mission;
        }
        return { ...mission, joined: false };
      });
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchMissions.fulfilled, (state, action) => action.payload);
  },
});

export const { leavingMission } = missionsSlice.actions;

export const selectMissions = (state) => state.missions;

export default missionsSlice.reducer;
