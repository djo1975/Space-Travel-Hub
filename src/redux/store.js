import { configureStore } from '@reduxjs/toolkit';
import rocketReducer from './Rockets/rocketSlice';
import dragonReducer from './Dragons/dragonSlice';
import missionsReducer from './Missions/missionsSlice';

const store = configureStore({
  reducer: {
    rockets: rocketReducer,
    dragons: dragonReducer,
    missions: missionsReducer,
  },
});

export default store;
