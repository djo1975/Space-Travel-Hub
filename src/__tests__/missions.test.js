import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../components/missions/missions';
import store from '../redux/store';

describe('Missions component', () => {
  test('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
