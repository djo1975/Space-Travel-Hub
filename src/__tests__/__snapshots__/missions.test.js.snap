// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions component renders correctly 1`] = `
<section
  className="missions-section"
>
  <h2
    className="missions-title"
  >
    Missions
  </h2>
  <table
    className="table-container"
  >
    <thead
      className="mission-th"
    >
      <tr
        className="mission-tr"
      >
        <th
          className="column-title"
        >
          Missions
        </th>
        <th
          className="column-title"
        >
          Description
        </th>
        <th
          className="column-title"
        >
          Status
        </th>
        <th
          className="column-title"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody
      className="table-body"
    />
  </table>
</section>
`;

exports[`Missions renders correctly 1`] = `
<section
  className="missions-section"
>
  <h2
    className="missions-title"
  >
    Missions
  </h2>
  <table
    className="table-container"
  >
    <thead
      className="mission-th"
    >
      <tr
        className="mission-tr"
      >
        <th
          className="column-title"
        >
          Missions
        </th>
        <th
          className="column-title"
        >
          Description
        </th>
        <th
          className="column-title"
        >
          Status
        </th>
        <th
          className="column-title"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody
      className="table-body"
    />
  </table>
</section>
`;

exports[`missions reducer renders correctly 1`] = `
<section
  className="missions-section"
>
  <h2
    className="missions-title"
  >
    Missions
  </h2>
  <table
    className="table-container"
  >
    <thead
      className="mission-th"
    >
      <tr
        className="mission-tr"
      >
        <th
          className="column-title"
        >
          Missions
        </th>
        <th
          className="column-title"
        >
          Description
        </th>
        <th
          className="column-title"
        >
          Status
        </th>
        <th
          className="column-title"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody
      className="table-body"
    />
  </table>
</section>
`;

exports[`renders correctly 1`] = `
<section
  className="missions-section"
>
  <h2
    className="missions-title"
  >
    Missions
  </h2>
  <table
    className="table-container"
  >
    <thead
      className="mission-th"
    >
      <tr
        className="mission-tr"
      >
        <th
          className="column-title"
        >
          Missions
        </th>
        <th
          className="column-title"
        >
          Description
        </th>
        <th
          className="column-title"
        >
          Status
        </th>
        <th
          className="column-title"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody
      className="table-body"
    />
  </table>
</section>
`;
